icicle/appUtils/msm/msm.cu

```diff
@@ -31,30 +31,71 @@ namespace msm {
 
     unsigned get_optimal_c(int bitsize) { return max((unsigned)ceil(log2(bitsize)) - 4, 1U); }
 
+    template <typename E>
+    __global__ void normalize_kernel(E* inout, E factor, int n)
+    {
+      int tid = blockIdx.x * blockDim.x + threadIdx.x;
+      if (tid < n) inout[tid] = (inout[tid] + factor - 1) / factor;
+    }
+
+    template <typename P>
+    __global__ void initialize_large_bucket_indices(
+      unsigned* sorted_bucket_sizes_sum,
+      unsigned nof_pts_per_thread,
+      unsigned nof_large_buckets,
+      unsigned log_nof_large_buckets,
+      unsigned* bucket_indices)
+    {
+      const int tid = blockIdx.x * blockDim.x + threadIdx.x;
+      if (tid >= nof_large_buckets) { return; }
+      unsigned start = (sorted_bucket_sizes_sum[tid] + nof_pts_per_thread - 1) / nof_pts_per_thread + tid;
+      unsigned end = (sorted_bucket_sizes_sum[tid + 1] + nof_pts_per_thread - 1) / nof_pts_per_thread + tid + 1;
+      for (unsigned i = start; i < end; i++) {
+        bucket_indices[i] = tid | ((i - start) << log_nof_large_buckets);
```

Comment:
are you sure this is correct when nof_buckets is not a power of two?

Reply:
It should be correct for non-powers-of-2 and in most cases `nof_large_buckets` is not a power of 2
What this line does is just packing two values into one number `bucket_indices[i]`. `tid` goes into the lowest `log_nof_large_buckets` bits and `i - start` goes into the rest. So `log_nof_large_buckets` in this case just means the number of bits needed to represent `tid` which varies from 0 to `nof_large_buckets`

Reply:
right, as long as log_nof_large_buckets = ceil(log2(nof_large_buckets)) this is correct. Since it's not verified inside the kernel, it could mix the two fields so that's why I suggested to make verify it at kernel launch.

Reply:
In principle we can just take log inside the kernel, I just wanted to avoid all threads doing identical work. Verification is definitely cheaper though, can do it inside the kernel if you want

Reply:
My motivation is to avoid debugging cases like nof_buckets=100, (int)log2(nof_buckets)=6.

I would personally add a comment about this assumption next to the kernel param and make sure the host is calling correctly. If you prefer assert inside the kernel it's fine too.

Reply:
Added a comment though maybe I should've done a more full doc comment for each kernel (but they are internal so I don't feel the need to spend too much time documenting them tbh)